singularity-eos/closure/mixed_cell_models.hpp

```diff
@@ -412,6 +421,60 @@ class PTESolverBase {
 
 } // namespace mix_impl
 
+template <typename EOSIndexer, typename RealIndexer>
+PORTABLE_INLINE_FUNCTION Real ApproxTemperatureFromRhoMatU(
+    const int nmat, EOSIndexer &&eos, const Real u_tot, RealIndexer &&rho,
+    RealIndexer &&vfrac, const Real Tguess = 0.0) {
+  // given material microphysical densities, volume fractions, and a total internal energy
+  // density (rho e => erg/cm^3), solve for the temperature that gives the right sum
+  // of material energies.  this should only be used for a rough guess since it has a
+  // hard coded and fairly loose tolerance
+  auto ufunc = [&](const Real T) {
+    Real usum = 0.0;
+    for (int m = 0; m < nmat; m++) {
+      usum += rho[m] * vfrac[m] * eos[m].InternalEnergyFromDensityTemperature(rho[m], T);
+    }
+    return usum;
+  };
+
+  Real ulo = ufunc(mix_params::minimum_temperature);
+  if (u_tot < ulo) return mix_params::minimum_temperature;
+  Real uhi = ufunc(mix_params::maximum_temperature);
+  if (u_tot > uhi) return mix_params::maximum_temperature;
+  Real lTlo = FastMath::lg(mix_params::minimum_temperature);
+  Real lThi = FastMath::lg(mix_params::maximum_temperature);
+  if (Tguess > mix_params::minimum_temperature &&
+      Tguess < mix_params::maximum_temperature) {
+    const Real ug = ufunc(Tguess);
+    if (ug < u_tot) {
+      lTlo = FastMath::lg(Tguess);
+      ulo = ug;
+    } else {
+      lThi = FastMath::lg(Tguess);
+      uhi = ug;
+    }
+  }
+  int iter = 0;
+  constexpr int max_iter = 10;
+  while (lThi - lTlo > 0.01 && iter < max_iter) {
+    // apply bisection which is much better behaved
+    // for materials that have a flat sie at low temperatures
+    const Real lT = 0.5 * (lTlo + lThi);
+    const Real uT = ufunc(FastMath::pow2(lT));
+    if (uT < u_tot) {
+      lTlo = lT;
+      ulo = uT;
+    } else {
+      lThi = lT;
+      uhi = uT;
+    }
+    iter++;
+  }
```

Comment:
Don't think this is necessary, but don't we have a pre-packaged bisection in the root finders file you could use?

Reply:
maybe.  just figured it was like 10 lines of code.  also, can `max_iter` be easily controlled via the function interface because that needs to be small here, but shouldn't typically be small for a bisection based root find.

Reply:
yeah fair enough.